# ECMAScript 2015



## [arrow-body-style](http://eslint.org/docs/rules/arrow-body-style)

- enforce or disallow the use of braces around arrow function body
  - enforces no braces where they can be omitted

``````javascript
{ 'arrow-body-style': 'error' }
``````

``````javascript
/** CORRECT USAGE */

const foo = baz => baz;
const bar = baz => {
  baz.qux();
  return baz;
}


/** INCORRECT USAGE */

const foo = val => { return val; };
``````



## [arrow-parens](http://eslint.org/docs/rules/arrow-parens)

- require parentheses around arrow function arguments

``````javascript
{ 'arrow-parens': 'off' }
``````



## [arrow-spacing](http://eslint.org/docs/rules/arrow-spacing)

- enforce consistent spacing before and after the arrow in arrow functions
  - requires one or more spaces before and after `=>` in an arrow function

``````javascript
{ 'arrow-spacing': 'error' }
``````

``````javascript
/** CORRECT USAGE */

const foo = baz => baz;


/** INCORRECT USAGE */

const foo = baz=>baz;
``````



## [constructor-super](http://eslint.org/docs/rules/constructor-super)

- require `super()` calls in constructors

``````javascript
{ 'constructor-super': 'error' }
``````

``````javascript
/** CORRECT USAGE */

class Bar extends Foo {
  constructor () {
    super();
    this.baz = true;
  }
}


/** INCORRECT USAGE */

// extends a class but does not call super() in its constructor
class Bar extends Foo {
  constructor () {
    this.baz = true;
  }
}

// calls super() but does not extend a class
class Foo {
  constructor () {
    super();
  }
}
``````



## [generator-star-spacing](http://eslint.org/docs/rules/generator-star-spacing)

- enforce consistent spacing around `*` operators in generator functions
  - require spacing between the `*` and the function name (or the opening parenthesis for anonymous generator functions)
  - disallow spacing between the `*` and the `function` keyword

``````javascript
{ 'generator-star-spacing': ['error', 'after'] }
``````

``````javascript
/** CORRECT USAGE */

function* foo() {}
const bar = function* () {};
const baz = { * qux () {} };


/** INCORRECT USAGE */

function *foo() {}
const bar = function *() {};
const baz = { *qux () {} };
``````



## [no-class-assign](http://eslint.org/docs/rules/no-class-assign)

- disallow reassigning class members

``````javascript
{ 'no-class-assign': 'error' }
``````

``````javascript
/** CORRECT USAGE */

let a = class {};
a = false;


/** INCORRECT USAGE */

class Foo {}
Foo = false;
``````



## [no-confusing-arrow](http://eslint.org/docs/rules/no-confusing-arrow)

- disallow arrow functions where they could be confused with comparisons

``````javascript
{ 'no-confusing-arrow': 'off' }
``````



## [no-const-assign](http://eslint.org/docs/rules/no-const-assign)

- disallow reassigning `const` variables

``````javascript
{ 'no-const-assign': 'error' }
``````

``````javascript
/** CORRECT USAGE */

let foo = true;
foo = false;

const bar = { baz: true };
bar.baz = false;


/** INCORRECT USAGE */

const foo = true;
foo = false;
``````



## [no-dupe-class-members](http://eslint.org/docs/rules/no-dupe-class-members)

- disallow duplicate class members

``````javascript
{ 'no-dupe-class-members': 'error' }
``````

``````javascript
/** CORRECT USAGE */

class Foo {
  bar () {}
  baz () {}
}


/** INCORRECT USAGE */

class Foo {
  bar () {}
  bar () {}
}
``````



## [no-duplicate-imports](http://eslint.org/docs/rules/no-duplicate-imports)

- disallow duplicate module imports

``````javascript
{ 'no-duplicate-imports': 'error' }
``````

``````javascript
/** CORRECT USAGE */

import { foo, bar } from 'foo';


/** INCORRECT USAGE */

import { foo } from 'foo';
import { bar } from 'foo';
``````



## [no-new-symbol](http://eslint.org/docs/rules/no-new-symbol)

- disallow `new` operators with the `Symbol` object

``````javascript
{ 'no-new-symbol': 'error' }
``````

``````javascript
/** CORRECT USAGE */

const foo = Symbol('foo');


/** INCORRECT USAGE */

const foo = new Symbol('foo');
``````



## [no-restricted-imports](http://eslint.org/docs/rules/no-restricted-imports)

- disallow specified modules when loaded by `import`

``````javascript
{ 'no-restricted-imports': 'off' }
``````



## [no-this-before-super](http://eslint.org/docs/rules/no-this-before-super)

- disallow `this`/`super` before calling `super()` in constructors

``````javascript
{ 'no-this-before-super': 'error' }
``````

``````javascript
/** CORRECT USAGE */

class Foo extends Bar {
  constructor () {
    super();
    super.qux = false;
    this.baz = false;
  }
}


/** INCORRECT USAGE */

class Foo extends Bar {
  constructor () {
    this.baz = false;
    super.qux = false;
    super();
  }
}
``````



## [no-useless-computed-key](http://eslint.org/docs/rules/no-useless-computed-key)

- disallow unnecessary computed property keys in object literals

``````javascript
{ 'no-useless-computed-key': 'error' }
``````

``````javascript
/** CORRECT USAGE */

const foo = {
  bar: true,
  baz: false
};


/** INCORRECT USAGE */

const foo = {
  ['bar']: true,
  baz: false
};
``````



## [no-useless-constructor](http://eslint.org/docs/rules/no-useless-constructor)

- disallow unnecessary constructors

``````javascript
{ 'no-useless-constructor': 'error' }
``````

``````javascript
/** CORRECT USAGE */

// class uses default ES2015 constructor
class Foo {} 

// class uses constructor with code
class Bar {
  constructor () {
    baz();
  }
}

/** INCORRECT USAGE */

// class uses empty constructor
class Foo {
  constructor () {}
}
``````



## [no-useless-rename](http://eslint.org/docs/rules/no-useless-rename)

- disallow renaming import, export, and destructured assignments to the same name

``````javascript
{ 'no-useless-rename': 'error' }
``````

``````javascript
/** CORRECT USAGE */

import * as foo from 'foo';
import { foo } from 'bar';
import { foo as bar } from 'baz';

export { foo };
export { foo as bar };
export { foo as bar } from 'foo';

const { foo } = bar;
const { foo: bar } = baz;
const { [foo]: foo } = bar;

function foo({ bar }) {}
function foo({ bar: baz }) {}

({ foo }) => {}
({ foo: bar }) => {}


/** INCORRECT USAGE */

import { foo as foo } from 'bar';
export { foo as foo };
export { foo as foo } from 'bar';
const { foo: foo } = bar;
const { 'foo': foo } = bar;
function foo({ bar: bar }) {}
({ foo: foo }) => {}
``````



## [no-var](http://eslint.org/docs/rules/no-var)

- require `let` or `const` instead of `var`

``````javascript
{ 'no-var': 'error' }
``````

``````javascript
/** CORRECT USAGE */

const foo = true;
let bar = false;


/** INCORRECT USAGE */

var foo = true;
``````



## [object-shorthand](http://eslint.org/docs/rules/object-shorthand)

- require or disallow method and property shorthand syntax for object literals

``````javascript
{ 'object-shorthand': 'off' }
``````



## [prefer-arrow-callback](http://eslint.org/docs/rules/prefer-arrow-callback)

- require arrow functions as callbacks

``````Javascript
{ 'prefer-arrow-callback': 'off' }
``````



## [prefer-const](http://eslint.org/docs/rules/prefer-const)

- require `const` declarations for variables that are never reassigned after declared

``````javascript
{ 'prefer-const': 'error' }
``````

``````javascript
/** CORRECT USAGE */

// never reassigned
const foo = true;

// reassigned after initialization
let bar = false;
bar = true;

// initialized after declaration
let baz;
function qux () {
  baz = false;
}


/** INCORRECT USAGE */

// never reassigned
let foo = true;
``````



## [prefer-numeric-literals](http://eslint.org/docs/rules/prefer-numeric-literals)

- disallow `parseInt()` in favor of binary, octal, and hexadecimal literals

``````javascript
{ 'prefer-numeric-literals': 'off' }
``````



## [prefer-reflect](http://eslint.org/docs/rules/prefer-reflect)

- require `Reflect` methods where applicable

``````javascript
{ 'prefer-reflect': 'off' }
``````



## [prefer-rest-params](http://eslint.org/docs/rules/prefer-rest-params)

- require rest parameters instead of `arguments`

``````javascript
{ 'prefer-rest-params': 'error' }
``````

``````javascript
/** CORRECT USAGE */

function foo (action, ...args) {
  action(...args);
}


/** INCORRECT USAGE */

function foo (action) {
  var args = [].slice.call(arguments, 1);
  action.apply(null, args);
}
``````



## [prefer-spread](http://eslint.org/docs/rules/prefer-spread)

- require spread operators instead of `.apply()`

``````javascript
{ 'prefer-spread': 'error' }
``````

``````javascript
/** CORRECT USAGE */

foo(...args);
foo.apply(obj, args); // distinct `this` binding is allowed


/** INCORRECT USAGE */

foo.apply(undefined, args);
foo.apply(null, args);
obj.foo.apply(obj, args);
``````



## [prefer-template](http://eslint.org/docs/rules/prefer-template)

- require template literals instead of string concatenation

``````javascript
{ 'prefer-template': 'error' }
``````

``````javascript
/** CORRECT USAGE */

const foo = `Hello, ${name}!`;
const bar = `Time: ${(12 * 60 * 60 * 1000)}`;


/** INCORRECT USAGE */

const foo = 'Hello, ' + name + '!';
const bar = 'Time: ' + (12 * 60 * 60 * 1000);
``````



## [require-yield](http://eslint.org/docs/rules/require-yield)

- require generator functions to contain `yield`

``````javascript
{ 'require-yield': 'error' }
``````

``````javascript
/** CORRECT USAGE */

function* foo () {
  yield 5;
  return 10;
}

function foo () {
  return 10;
}


/** INCORRECT USAGE */

function* foo () {
  return 10;
}
``````



## [rest-spread-spacing](http://eslint.org/docs/rules/rest-spread-spacing)

- enforce spacing between rest and spread operators and their expressions
  - disallows whitespace between spread operators and their expressions

``````javascript
{ 'rest-spread-spacing': 'error' }
``````

``````javascript
/** CORRECT USAGE */

foo(...args);
function foo (...args) {}


/** INCORRECT USAGE */

foo(... args);
function foo (... args) {}
``````



## [sort-imports](http://eslint.org/docs/rules/sort-imports)

- enforce sorted import declarations within modules

``````javascript
{ 'sort-imports': 'off' }
``````



## [symbol-description](http://eslint.org/docs/rules/symbol-description)

- require symbol descriptions

``````javascript
{ 'symbol-description': 'off' }
``````



## [template-curly-spacing](http://eslint.org/docs/rules/template-curly-spacing)

- require or disallow spacing around embedded expressions of template strings

``````javascript
{ 'template-curly-spacing': 'error' }
``````

``````javascript
/** CORRECT USAGE */

`hello, ${people.name}!`;
`hello, ${
  people.name
}!`;


/** INCORRECT USAGE */

`hello, ${ people.name}!`;
`hello, ${people.name }!`;
`hello, ${ people.name }!`;
``````



## [yield-star-spacing](http://eslint.org/docs/rules/yield-star-spacing)

- require or disallow spacing around the `*` in `yield*` expressions

``````javascript
{ 'yield-star-spacing': ['error', 'after'] }
``````

``````javascript
/** CORRECT USAGE */

function* generator () {
  yield* other();
}


/** INCORRECT USAGE */

function* generator () {
  yield * other();
}

function* generator () {
  yield *other();
}
``````

